File: aoc2022/kotlin/src/main/kotlin/Day03.kt
import util.AdventOfCode

object Day03 : AdventOfCode() {
    fun Char.priority() =
        this.code - if (this.isLowerCase()) 96 else 38

    fun part1(input: Sequence<String>) =
        input.flatMap {
            it.substring(0, it.length / 2).toSet().intersect(it.substring(it.length / 2).toSet())
        }.sumOf { it.priority() }

    fun part2(input: Sequence<String>) =
        input.map { it.toSet() }
            .chunked(3) { it.reduce(Set<Char>::intersect) }
            .flatten()
            .sumOf { it.priority() }

    @JvmStatic
    fun main(args: Array<String>) {
        println("Day 3")
        println("\tPart 1: ${part1(inputFileLineSequence)}")
        println("\tPart 2: ${part2(inputFileLineSequence)}")
    }
}

File: aoc2022/fsharp/Days/Day03.fs
module Day03

let lowercaseMin = 'a' |> int
let lowercaseMax = 'z' |> int


let priority (inputChar: char) =
    match (inputChar |> int) with
    | c when c >= lowercaseMin && c <= lowercaseMax -> c - 96
    | c -> c - 38

let intersect (line: string) =
    let first = line.Substring(0, line.Length / 2) |> set
    let second = line.Substring(line.Length / 2) |> set
    first |> Set.intersect <| second

let part1 (input: string seq) =
    input |> Seq.collect intersect |> Seq.sumBy priority

let part2 (input: string seq) =
    input
    |> Seq.map set
    |> Seq.chunkBySize 3
    |> Seq.collect (Seq.reduce Set.intersect)
    |> Seq.sumBy priority
